src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -75,8 +75,49 @@ public ProxyWhitelist(Collection<? extends Whitelist> delegates) {
         reset(delegates);
     }
 
+    /**
+     * This should only be called from {@link ProxyWhitelist#reset(Collection)}. So we can assume that a write 
+     * lock is held.
+     * @param wrapper the wrapper
+     */
+    private void addWrapper(ProxyWhitelist wrapper) {
+        assert wrapper.lock.writeLock().isHeldByCurrentThread();
+        lock.writeLock().lock();
+        try {
+            wrappers.put(wrapper, null);
+            lock.readLock().lock();
+        } finally {
+            lock.writeLock().unlock();
```

Comment:
This is now officially over my head! https://en.wikipedia.org/wiki/Software_transactional_memory where are you

Reply:
@jglick 

Downgrading `this.lock` from write lock to read lock to reduce contention? Do we need to comment it? 
```suggestion
            // The rest of the method only reads the current instance's fields
            // So we downgrade this.lock from write locked to read locked to reduce contention
            lock.readLock().lock();
        } finally {
            lock.writeLock().unlock();
```

Would it be better to extend @dwnusbaum 's feedback about taking write lock to this? Then we'd eagerly take write lock on each wrapper and keep it since it will be taken later when calling `reset()` on [lines 170-172](https://github.com/jenkinsci/script-security-plugin/pull/356/files#diff-b466c3e3e537c5a2de00bc55bcfa18465a442ea37ac3d92a7f2cd9ded4221f7dR170-R172).